🐛 Bug: RuleTester report errors in wrong test cases

### Bug Report Checklist

- [x] I am using the latest published version of all Flint packages.
- [x] I have [searched for related issues](https://github.com/flint-fyi/flint/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

`RuleTester` tests should be completely isolated. 

### Actual

Splitting out of #2398 -> #2435: there's some kind of odd internal state shenanigan where `RuleTester` tests are surfacing reports in the wrong case. This results in a lot of test failures locally. 

<details>
<summary>Using `pnpm run test run alert` as an example</summary>

```plaintext
josh ~/repos/flint $ pnpm run test run alert

> @flint/workspace@0.12.0 test /Users/josh/repos/flint
> vitest --typecheck run alert

 RUN  v4.0.18 /Users/josh/repos/flint

 ❯  browser  src/rules/alerts.test.ts (8 tests | 7 failed) 572ms
   ❯ alerts (8)
     ❯ invalid (4)
       ✓ 
alert("...");
  558ms
       × 
window.alert("...");
 5ms
       × 
globalThis.confirm("...");
 2ms
       × 
self.confirm("...");
 1ms
     ❯ valid (4)
       × console.log("..."); 1ms
       × window.console.log("x"); 1ms
       × other.alert(); 1ms
       × 
                        declare function alert(message: string): void;
                        alert("...");
                        export {};
                 1ms

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 7 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > invalid > 
window.alert("...");

AssertionError: Expected values to be strictly equal:
+ actual - expected

  '\n' +
    'window.alert("...");\n' +
+   '~~~~~\n' +
+   'The global `alert()` API blocks the main thread and interrupts users.\n'
-   '       ~~~~~\n' +
-   '       The global `alert()` API blocks the main thread and interrupts users.\n'


- Expected
+ Received


  window.alert("...");
-        ~~~~~
-        The global `alert()` API blocks the main thread and interrupts users.
+ ~~~~~
+ The global `alert()` API blocks the main thread and interrupts users.


 ❯ ../rule-tester/src/RuleTester.ts:165:11
    163|    const actualSnapshot = createReportSnapshot(testCase.code, reports);
    164| 
    165|    assert.equal(actualSnapshot, testCase.snapshot);
       |           ^
    166| 
    167|    const actualOutput = createOutput(reports, testCaseNormalized);

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > invalid > 
globalThis.confirm("...");

AssertionError: Expected values to be strictly equal:
+ actual - expected

  '\n' +
    'globalThis.confirm("...");\n' +
+   '~~~~~\n' +
+   'The global `alert()` API blocks the main thread and interrupts users.\n'
-   '           ~~~~~~~\n' +
-   '           The global `confirm()` API blocks the main thread and interrupts users.\n'


- Expected
+ Received


  globalThis.confirm("...");
-            ~~~~~~~
-            The global `confirm()` API blocks the main thread and interrupts users.
+ ~~~~~
+ The global `alert()` API blocks the main thread and interrupts users.


 ❯ ../rule-tester/src/RuleTester.ts:165:11
    163|    const actualSnapshot = createReportSnapshot(testCase.code, reports);
    164| 
    165|    assert.equal(actualSnapshot, testCase.snapshot);
       |           ^
    166| 
    167|    const actualOutput = createOutput(reports, testCaseNormalized);

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > invalid > 
self.confirm("...");

AssertionError: Expected values to be strictly equal:
+ actual - expected

  '\n' +
    'self.confirm("...");\n' +
+   '~~~~~\n' +
+   'The global `alert()` API blocks the main thread and interrupts users.\n'
-   '     ~~~~~~~\n' +
-   '     The global `confirm()` API blocks the main thread and interrupts users.\n'


- Expected
+ Received


  self.confirm("...");
-      ~~~~~~~
-      The global `confirm()` API blocks the main thread and interrupts users.
+ ~~~~~
+ The global `alert()` API blocks the main thread and interrupts users.


 ❯ ../rule-tester/src/RuleTester.ts:165:11
    163|    const actualSnapshot = createReportSnapshot(testCase.code, reports);
    164| 
    165|    assert.equal(actualSnapshot, testCase.snapshot);
       |           ^
    166| 
    167|    const actualOutput = createOutput(reports, testCaseNormalized);

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > valid > console.log("...");
AssertionError: Expected values to be strictly deep-equal:
+ actual - expected

+ 'console.log("...");\n' +
+   ' ~~~~~\n' +
+   ' The global `alert()` API blocks the main thread and interrupts users.'
- 'console.log("...");'


- Expected
+ Received

  console.log("...");
+  ~~~~~
+  The global `alert()` API blocks the main thread and interrupts users.

 ❯ ../rule-tester/src/RuleTester.ts:238:12
    236| 
    237|    if (reports.length) {
    238|     assert.deepStrictEqual(
       |            ^
    239|      createReportSnapshot(testCaseNormalized.code, reports),
    240|      testCaseNormalized.code,

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[4/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > valid > window.console.log("x");
AssertionError: Expected values to be strictly deep-equal:
+ actual - expected

+ 'window.console.log("x");\n' +
+   ' ~~~~~\n' +
+   ' The global `alert()` API blocks the main thread and interrupts users.'
- 'window.console.log("x");'


- Expected
+ Received

  window.console.log("x");
+  ~~~~~
+  The global `alert()` API blocks the main thread and interrupts users.

 ❯ ../rule-tester/src/RuleTester.ts:238:12
    236| 
    237|    if (reports.length) {
    238|     assert.deepStrictEqual(
       |            ^
    239|      createReportSnapshot(testCaseNormalized.code, reports),
    240|      testCaseNormalized.code,

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[5/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > valid > other.alert();
AssertionError: Expected values to be strictly deep-equal:
+ actual - expected

+ 'other.alert();\n' +
+   ' ~~~~~\n' +
+   ' The global `alert()` API blocks the main thread and interrupts users.'
- 'other.alert();'


- Expected
+ Received

  other.alert();
+  ~~~~~
+  The global `alert()` API blocks the main thread and interrupts users.

 ❯ ../rule-tester/src/RuleTester.ts:238:12
    236| 
    237|    if (reports.length) {
    238|     assert.deepStrictEqual(
       |            ^
    239|      createReportSnapshot(testCaseNormalized.code, reports),
    240|      testCaseNormalized.code,

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[6/7]⎯

 FAIL   browser  src/rules/alerts.test.ts > alerts > valid > 
                        declare function alert(message: string): void;
                        alert("...");
                        export {};

AssertionError: Expected values to be strictly deep-equal:
+ actual - expected

  '\n' +
    '\t\t\tdeclare function alert(message: string): void;\n' +
+   '\t\t\t~~\n' +
+   '\t\t\tThe global `alert()` API blocks the main thread and interrupts users.\n' +
    '\t\t\talert("...");\n' +
    '\t\t\texport {};\n' +
    '\t\t'


- Expected
+ Received


                        declare function alert(message: string): void;
+                       ~~
+                       The global `alert()` API blocks the main thread and interrupts users.
                        alert("...");
                        export {};
  

 ❯ ../rule-tester/src/RuleTester.ts:238:12
    236| 
    237|    if (reports.length) {
    238|     assert.deepStrictEqual(
       |            ^
    239|      createReportSnapshot(testCaseNormalized.code, reports),
    240|      testCaseNormalized.code,

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[7/7]⎯


 Test Files  1 failed (1)
      Tests  7 failed | 1 passed (8)
Type Errors  no errors
   Start at  12:26:19
   Duration  2.35s (transform 802ms, setup 74ms, import 1.54s, tests 572ms, environment 0ms)

 ELIFECYCLE  Test failed. See above for more details.
```

</details>

However, if you only use `{ only: true }` on any of the individual tests, they still pass. So this is a bug in `RuleTester`.

### Additional Info

I bisected this in history with `pnpm run test run wrapperObjects`:

* 9f4d531f (#2321): has the failure ❌ 
* 13983caa: does not have the failure

Interestingly, this doesn't show up in CI. #2435 has a clean build. Which kind of does make sense given that this is something to do with string casing and normalization. 

❤️‍🔥

--- a/packages/core/src/host/createVFSLinterHost.ts
+++ b/packages/core/src/host/createVFSLinterHost.ts
@@ -61,7 +61,7 @@ export function createVFSLinterHost(
 
 	interface VfsFile {
 		content: string;
-		path: PathKey;
+		path: string;
 	}
 
 	const fileMap = new Map<PathKey, VfsFile>();
@@ -73,10 +73,11 @@ export function createVFSLinterHost(
 	>();
 
 	function watchEvent(
-		normalizedFilePathAbsolute: PathKey,
+		normalizedFilePathAbsolute: string,
 		fileEvent: LinterHostFileWatcherEvent,
 	) {
-		for (const watcher of fileWatchers.get(normalizedFilePathAbsolute) ?? []) {
+		const fileKey = pathKey(normalizedFilePathAbsolute, caseSensitiveFS);
+		for (const watcher of fileWatchers.get(fileKey) ?? []) {
 			watcher(fileEvent);
 		}
 
@@ -108,7 +109,6 @@ export function createVFSLinterHost(
 	}
 	const host: VFSLinterHost = {
 		fileTypeSync(pathAbsolute) {
-			pathAbsolute = normalizePath(pathAbsolute);
 			const key = pathKey(pathAbsolute, caseSensitiveFS);
 			const keySlash = dirnameKey(pathAbsolute, caseSensitiveFS);
 			for (const fileKey of fileMap.keys()) {
@@ -183,8 +183,8 @@ export function createVFSLinterHost(
 			return host.readFileSync(filePathAbsolute);
 		},
 		readFileSync(filePathAbsolute) {
-			filePathAbsolute = normalizePath(filePathAbsolute);
-			const file = fileMap.get(pathKey(filePathAbsolute, caseSensitiveFS));
+			const key = pathKey(filePathAbsolute, caseSensitiveFS);
+			const file = fileMap.get(key);
 			if (file != null) {
 				return file.content;
 			}
@@ -194,7 +194,6 @@ export function createVFSLinterHost(
 			return undefined;
 		},
 		vfsDeleteFile(filePathAbsolute) {
-			filePathAbsolute = normalizePath(filePathAbsolute);
 			const key = pathKey(filePathAbsolute, caseSensitiveFS);
 			const file = fileMap.get(key);
 			if (file == null) {
@@ -207,17 +206,14 @@ export function createVFSLinterHost(
 			return new Map(Array.from(fileMap.values(), (f) => [f.path, f.content]));
 		},
 		vfsUpsertFile(filePathAbsolute, content) {
-			filePathAbsolute = normalizePath(filePathAbsolute);
 			const key = pathKey(filePathAbsolute, caseSensitiveFS);
 			const existing = fileMap.get(key);
-			// TODO: Thread PathKey through the rest of the core.
-			const storedPath = existing?.path ?? (filePathAbsolute as PathKey);
+			const storedPath = existing?.path ?? normalizePath(filePathAbsolute);
 			const fileEvent = existing != null ? "changed" : "created";
 			fileMap.set(key, { content, path: storedPath });
 			watchEvent(storedPath, fileEvent);
 		},
 		watchDirectorySync(directoryPathAbsolute, callback, options) {
-			directoryPathAbsolute = normalizePath(directoryPathAbsolute);
 			const key = pathKey(directoryPathAbsolute, caseSensitiveFS);
 			const collection = options.recursive
 				? recursiveDirectoryWatchers
@@ -244,7 +240,6 @@ export function createVFSLinterHost(
 			};
 		},
 		watchFileSync(filePathAbsolute, callback, options) {
-			filePathAbsolute = normalizePath(filePathAbsolute);
 			const key = pathKey(filePathAbsolute, caseSensitiveFS);
 			let watchers = fileWatchers.get(key);
 

